### PR TITLE
Do not allow custom change types to execute twice DAT-14051

### LINKED ISSUE #3945: executeCommand runs twice per changeSet
### Search first

- [X] I searched and no similar issues were found

### Description

The `executeCommand` is being executed twice per changeSet, since Liquibase 4.19.1. This was working as expected on Liquibase 4.19.0.

We can see that the [ExecuteShellCommandChange](https://github.com/liquibase/liquibase/blob/63f445f85beda72e7aafc8bb18340713049ecc5c/liquibase-core/src/main/java/liquibase/change/core/ExecuteShellCommandChange.java#L122) uses the generateStatements function to execute the shell command conditionally. This seems to be now called [while reading the changesets](https://github.com/liquibase/liquibase/blob/63f445f85beda72e7aafc8bb18340713049ecc5c/liquibase-core/src/main/java/liquibase/changelog/ChangeSet.java#L714), which will effectively execute the command prematurely.

A git bisect tracked the commit that introduced this erratic behavior to 005857b1aa387896acf9a6bb12a813fbb34281a0

### Steps To Reproduce

Run the following changeSet:

```xml
<changeSet id="shell command" author="joserebelo">
    <executeCommand executable="/bin/bash">
        <arg value="-c"/>
        <arg value="echo hi >> /tmp/liquibase_shell_out"/>
    </executeCommand>
</changeSet>
```

Notice that `/tmp/liquibase_shell_out` will actually contains 2 lines, since the executeCommand change ran twice.

### Expected/Desired Behavior

The executeCommand change should only be executed once per changeSet.

### Liquibase Version

4.19.1, also confirmed on master (fa6ecb783a6931cddce5f07773f40217b6c9d7c0)

### Database Vendor & Version

PostgreSQL 9.6

### Liquibase Integration

cli

### Liquibase Extensions

_No response_

### OS and/or Infrastructure Type/Provider

Ubuntu

### Additional Context

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR (Thank you!)

### PR DESCRIPTION
**Pull request by wwillard7800**
## Impact

- [x] Bug fix (non-breaking change which fixes expected existing functionality)
- [ ] Enhancement/New feature (adds functionality without impacting existing logic)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 
## Description
The addition of context data to logging was causing custom changes to be executed twice.  These changes prevent that.

Fixes #3945

### CODE REVIEW

File: liquibase-core/src/main/java/liquibase/change/Change.java [hunk 1]
```diff
@@ -140,4 +142,16 @@ public interface Change extends LiquibaseSerializable, Plugin, ExtensibleObject
...
+
+    /**
+     *
+     * Override this to allow this instance be in non-execute mode
+     * Default return is false
+     *
+     * @return   boolean       True if non-execute mode
+     *
+     */
+    default boolean allowNonExecuteMode() {
```
Comment (filipelautert) [inline liquibase-core/src/main/java/liquibase/change/Change.java:154]: This method is not used anymore right?

File: liquibase-core/src/main/java/liquibase/change/core/ExecuteShellCommandChange.java [hunk 2]
```diff
@@ -378,6 +370,11 @@ protected void customLoadLogic(ParsedNode parsedNode, ResourceAccessor resourceA
         }
     }
 
+    @Override
```
Comment (filipelautert) [inline liquibase-core/src/main/java/liquibase/change/core/ExecuteShellCommandChange.java:373]: Here too?

File: liquibase-core/src/main/java/liquibase/change/core/ExecuteShellCommandChange.java [hunk 3]
```diff
@@ -154,12 +146,12 @@ public Sql[] generate(Database database) {
                         throw new UnexpectedLiquibaseException("Error executing command: " + e.getLocalizedMessage(), e);
                     }
 
-                    return null;
+                    return new Sql[0];
```
Comment (filipelautert) [inline liquibase-core/src/main/java/liquibase/change/core/ExecuteShellCommandChange.java:149]: Is this change is expected?

File: liquibase-core/src/main/java/liquibase/change/Change.java (hunk 1 above)
Comment (wwillard7800) [reply liquibase-core/src/main/java/liquibase/change/Change.java:154]: Correct.  I need to remove it.

File: liquibase-core/src/main/java/liquibase/change/core/ExecuteShellCommandChange.java (hunk 2 above)
Comment (wwillard7800) [reply liquibase-core/src/main/java/liquibase/change/core/ExecuteShellCommandChange.java:373]: Same

File: liquibase-core/src/main/java/liquibase/change/core/ExecuteShellCommandChange.java (hunk 3 above)
Comment (wwillard7800) [reply liquibase-core/src/main/java/liquibase/change/core/ExecuteShellCommandChange.java:149]: I followed a suggestion made by Sonar, but I probably shouldn't have made that change.  I will revert it, just to be safe.

**Comment by XDelphiGrl** [comment]
Functional Tests Started Here: [https://github.com/liquibase/liquibase-pro-tests/actions/runs/4611625991](https://github.com/liquibase/liquibase-pro-tests/actions/runs/4611625991)

Test Harness Execution Started Here: [https://github.com/liquibase/liquibase-test-harness/actions/runs/4611629843](https://github.com/liquibase/liquibase-test-harness/actions/runs/4611629843)

**Comment by tarioch** [comment]
Is it possible that this broke the SQL Generation mode for custom sql changes? I'm still trying to track this down, but it looks like since this change we're no longer getting the SQL generated for the custom changes.

**Pull request merged**